[godot 4]

pcen opened the pull request:
Development <- godot4

pcen committed:
fix compiler warnings, .net 6.0

pcen committed:
wip upgrade to godot 4

pcen committed:
delete messaging metadata file

pcen committed:
removed .godot

pcen committed:
compiles, but commented out all references to 2d arrays since they don't compile in autogenerated godot code

pcen committed:
getting cannot load file or assembly error for serilog

pcen committed:
fix Serilog assembly not found by adding CopyLocalLockFileAssemblies to csproj

pcen committed:
CanvasItem Update -> QueueRedraw

pcen committed:
compile with Godot 4 beta 17

pcen committed:
re-connect unit buttons

pcen committed:
manually fix all signal connections being removed by project converter

pcen committed:
update sprite frames to godot 4 fields

pcen committed:
build with rc-1

pcen committed:
uncomment code that broke earlier 4.0 betas

pcen committed:
fix units not rendering at all - units render but look really weird, and some loose layer objects such as hills render with the wrong offset

pcen committed:
wip

pcen committed:
check in .import files to git as per godot docs

pcen committed:
fix terrain border lines by adding premultiplied blending mode

pcen committed:
wip removing custom shader and using godot animations

pcen committed:
separate animations into tint layer and regular layer

pcen committed:
Godot AnimatedSprite2D somewhat working for running warrior texture

pcen committed:
working unit animations for all types

pcen committed:
define animation key name functions

pcen committed:
fix unit render test scene

pcen committed:
load animations only once

pcen committed:
re-add vertical offset to unit animation based on frame size

pcen committed:
clean up some code, also figured out that city label text being white rectangles is due to premult alpha mode being enabled on the loose view

pcen committed:
use nearest filter by default

pcen committed:
rename text_entered signal to text_submitted

pcen committed:
set default texture filter to nearest for entire project instead of C7Game node

pcen committed:
fix unit animation test scene

pcen committed:
add shadow to ByteArrayWithTintToImage when loading flic frames

pcen committed:
load cursor as sprite instead of using custom shader

pcen committed:
target default c# language version for .net version

pcen committed:
fix unit right click menu position

pcen committed:
simplify how menu code works

pcen committed:
add nuget package for encoding

pcen committed:
add converter from stackoverflow for 2d array json serialization

pcen committed:
rebase on head of Development

pcen committed:
testing

pcen committed:
use godot actions instead of hardcoded keys

pcen committed:
fix messed up formatting

pcen committed:
remove debug prints

pcen committed:
replace GD.Print with log.Debug in Game.cs and add end_turn C7Action constant

pcen committed:
replace all action string literals with constants

pcen committed:
Merge pull request #405 from C7-Game/pcen/use-input-actions

Godot4: Use Godot Action System for Input

pcen committed:
main menu quit button works

pcen committed:
fixing ui layout

pcen committed:
layout

pcen committed:
fix ui elements being loaded from pcx with shadow

pcen committed:
ui fits nicely on screen

pcen committed:
remove debug print

pcen committed:
Merge pull request #410 from C7-Game/pcen/godot4-ui-fixes

Godot 4 UI Fixes

pcen committed:
fix popup bugs

pcen committed:
make type-safe, fix memory leak

pcen committed:
Revert "Godot 4 UI Fixes"

pcen committed:
Merge pull request #416 from C7-Game/revert-410-pcen/godot4-ui-fixes

Revert "Godot 4 UI Fixes"

pcen committed:
Revert "Revert "Godot 4 UI Fixes""

pcen committed:
Merge pull request #417 from C7-Game/revert-416-revert-410-pcen/godot4-ui-fixes

Revert "Revert "Godot 4 UI Fixes""

QuintillusCFC committed:
Sets the default font size to 14, and sets it to use Noto Sans.  14 was the default in Godot 3.x; this causes it to match existing functionality.  Downside, for some reason the overrides to font size for things like headers in dialogs aren't working; I verified that they weren't without these changes either.

Also, switch to Godot 4.1.

pcen committed:
incrementing id class and factory

pcen committed:
wip

pcen committed:
unit and tile IDs, having issues with updating save json when changing format

pcen committed:
remove debug print

pcen committed:
update save file to include tile IDs

pcen committed:
Merge pull request #418 from C7-Game/default-font-size-14

[4.x] Default font size 14

pcen committed:
Merge pull request #415 from C7-Game/pcen/type-safe-parameter-wrapper

[Godot 4] Make ParameterWrapper type safe and remove memory leak

pcen committed:
use ShowingPopup instead of Visible on PopupOverlay class

benskywalker committed:
Fixing Movement UI With Mixed Numbers (#429)

* Fixing Movement UI With Mixed Numbers

---------

Co-authored-by: benskywalker <benbowles2@gmail.com>

QuintillusCFC committed:
Merge pull request #414 from C7-Game/pcen/fix-popup-bugs

[Godot 4] Fix Popup Bugs

QuintillusCFC committed:
Two fixes:

1. City label being too wide.  This is fixed by specifying the size of the font used for the label so Godot can calculate it correctly.  (Note it doesn't use the cache, as otherwise it would change the size everywhere).
2. Remove the square box around the city graphics.  This was caused by using shadows when they shouldn't have been used; this is now configurable. (Not 100% sure why this only affected Godot 4)

QuintillusCFC committed:
Fix the pop size not having the right font size.

QuintillusCFC committed:
Two fixes:

1. Set the header size on popups.  Affects pop-ups such as Main Menu (Esc), Build City, Disband, and Error, all in one line.
2. Fix the font colors not being correct on hover or focus.  This was because Godot 4 changed the names from e.g. font_color_hover to font_hover_color, and we had the old names.

QuintillusCFC committed:
Fix the popup font size being slightly too small, and the button text being misaligned with the button.

QuintillusCFC committed:
Apply the same button/label font-size/alignment fixes to the Main Menu.

Also set the title to "C7 - Godot 4" so I can more easily tell which one I'm looking at, now that the most obvious differences are no longer present.

QuintillusCFC committed:
Fix shadows being turned on for the button graphics file.  Verified that the Godot 3.5 no-shadows version matches Civ3.

QuintillusCFC committed:
Turn off the pull-out sidebar and "End Turn" button from the super-early-pre-Aztec-alpha days.

QuintillusCFC committed:
Right-click font should be black, not orange.

QuintillusCFC committed:
Fix shadow issues with advisor pop-up graphics.

QuintillusCFC committed:
Highlighted build city text should be black, not white, as should the cursor (I-beam is still white, haven't figured out how to fix that yet).

QuintillusCFC committed:
Merge pull request #439 from C7-Game/godot-4-graphics-fixes

Fix Godot 4 City Graphics Regressions

pcen committed:
also search for steam copy of civ3 on windows

pcen committed:
Merge pull request #442 from C7-Game/search-for-steam-civ3-on-windows

also search for steam copy of civ3 on windows

pcen committed:
Merge pull request #440 from C7-Game/fix-popups-and-font-colors

Fix Popups and Font Colors

pcen committed:
make caret color black by default

pcen committed:
Merge pull request #446 from C7-Game/pcen/fix-caret-color

[godot 4] make caret color black by default

pcen committed:
prevent clicking other controls when popup is open

pcen committed:
restore old UI behavior for popups

pcen committed:
adjust spacing for ENTER or SPACEBAR... next turn message

pcen committed:
Merge pull request #447 from C7-Game/pcento/fix-popup-input-handled

Prevent mouse interactions with map and other UI elements when popup is open

pcen committed:
Save without references for godot4 branch (#443)

* save types without references

* document usage of BuildDevSave

* add tests, do not save which units are on tile in save

* do not store city on tile in save

* when loading tile knowledge from .sav into SaveGame, don't use AddTilesToKnown since neighbors shouldn't automatically be included

* add note about missing Civilizations in GameData

* pass round trip load save -> game data -> write save file bytes are the same

* clean up id

* more cleanup

* re-add Civ3ExtraInfo

* fixes to make save game load in godot4 branch without tilemap renderer

* clean up C7GameDataTests data directory and download input files

---------

Co-authored-by: Travis Christian <wildweazel@gmail.com>

pcen committed:
fix save game unit tests (#448)

* change line endings to LF

* don't download test SAV if it already exists locally

WildWeazel commented on C7GameDataTests/GameMapTest.cs:
```diff
@@ -0,0 +1,14 @@
+public class GameMapTest
+{
+	[Fact]
+	public void CityWith2ProductionPerTurn_ShouldReturn1TurnIf9_of_10ProductionDone()
```
not what this test does... copy-paste error?

WildWeazel commented on C7GameData/Save/TileLocation.cs:
```diff
@@ -0,0 +1,17 @@
+namespace C7GameData.Save {
+	public struct TileLocation {
+		public int x, y;
```
since vectors now use capitalized X and Y I'm thinking we should follow the same convention

WildWeazel replied:
#482

WildWeazel commented on C7Engine/C7Engine.csproj:
```diff
@@ -1,16 +1,15 @@
   <PropertyGroup>
-    <TargetFramework>net472</TargetFramework>
-    <LangVersion>8.0</LangVersion>
+    <TargetFramework>net7.0</TargetFramework>
```
since we're on 4.2 and .NET7 is already out of support by now, should we go ahead and target .NET8 (LTS version) and C# 10+?

WildWeazel replied:
#483

WildWeazel commented on C7/Civ3Map/Civ3Map.cs:
```diff
@@ -77,21 +78,21 @@ private void LoadTileSet(int fileID)
-				TS.TileSetTexture(id, Txtr);
-				TS.TileSetRegion(id, new Rect2(x, y, 128, 64));
-				TileIDLookup[fileID, (x / 128) + (y / 64) * (PcxTxtr.Width / 128)] = id;
+			for (int y = 0; y < PcxTxtr.Height; y += 64) {
+				for (int x = 0; x < PcxTxtr.Width; x+= 128/*, id++*/) {
+					// TS.Create
+					// TS.CreateTile(id);
+					// TS.TileSetTexture(id, Txtr);
+					// TS.TileSetRegion(id, new Rect2(x, y, 128, 64));
+					// TileIDLookup[fileID, (x / 128) + (y / 64) * (PcxTxtr.Width / 128)] = id;
```
is this entire block now useless?

WildWeazel commented on C7/Civ3Unit/Civ3Unit.cs:
```diff
@@ -2,177 +2,173 @@
-        }
-    }
+// This is just to add some movement to an AnimatedSprite2D
+public partial class MovingSprite : AnimatedSprite2D {
```
what does this movement do and where do the numbers come from? is it assuming something about window size?

WildWeazel commented on C7/PCXToGodot.cs:
```diff
@@ -95,21 +95,48 @@ private static Image ByteArrayToImage(byte[] colorIndices, byte[,] palette, int
+				baseLayer[i] = 0; // transparent
+			} else if (shadow) {
+				// shadow belongs to the base texture
+				baseLayer[i] = ((int)new Color(1.0f, 1.0f, 1.0f, (float)index - 224f / 239f - 224f).ToArgb32());
```
there's extra float math here that could be simplified to (float)(index-224)/15

also more constants for readability

WildWeazel commented on C7GameData/Tile.cs:
```diff
@@ -260,6 +259,20 @@ public static (int, int) toCoordDiff(this TileDirection dir)
 			}
 		}
+
+		// public static string shortName(this TileDirection dir) {
```
delete if really not useful, but I can see this being needed later

WildWeazel commented on C7/UIElements/Popups/PopupOverlay.cs:
```diff
@@ -1,83 +1,106 @@
+		return GetNode<Control>(ControlNodePath);
+	}
+
+	private void setMouseFilter(Node n, MouseFilterEnum filter) {
+		foreach (Node child in n?.GetChildren()) {
+			setMouseFilter(child, filter);
+		}
+		if (n is Control control) {
+			control.MouseFilter = filter;
+		}
+	}
```
let's make a TODO to look for a pattern for this. there has to be a better way than recursing through the whole scene

WildWeazel replied:
#452

TomWerner committed:
UnitButtons.cs: Only display buttons for available actions

This avoids the problem of the "road" button showing up when a worker is standing on a tile with a road.

Please let me know if there's a good way to test this.

TomWerner committed:
UnitButtons.cs: Only display buttons for available actions

This avoids the problem of the "road" button showing up when a worker is standing on a tile with a road.

Please let me know if there's a good way to test this.

TomWerner committed:
Right click menu: Add a "change production" option when right clicking on a city

This avoids the player having to know about the "shift+right click" shortcut to
change production.

Despite playing a fair bit of civ3, I didn't know about this shortcut and had to
figure it out by looking at the c7 code. These extra menu items make this basic
game functionality more discoverable for new players.

TomWerner committed:
Right click menu: Add a "change production" option when right clicking on a city

This avoids the player having to know about the "shift+right click" shortcut to
change production.

Despite playing a fair bit of civ3, I didn't know about this shortcut and had to
figure it out by looking at the c7 code. These extra menu items make this basic
game functionality more discoverable for new players.

TomWerner committed:
Fixed merge

TomWerner committed:
WIP on move cost estimates

WildWeazel committed:
Merge branch 'TomWerner-add-change-production-button' into godot4

TomWerner committed:
Boat pathing: Properly handle "goto" for water units (#456)

I noticed that while I could move a boat around with the arrow keys, I couldn't
move it with "G" or the "goto" button. Since my laptop doesn't have a numpad,
this made it tricky to explore.

I think this is a fairly straightforward fix to the problem:

1. I added an implementation of the EdgeWalker interface that handles water tiles,
   including cities in the set of neighbors to allow canal cities, and pathing
   back into coastal cities.

2. Require specifying land vs water units when picking a pathing algorithm.

I did some manual testing, including building a canal city north of the default
spawn, and confirmed I could path through it. I think building the city on a
hill amusingly resulted in the boat movement cost being 2, but besides that
it worked.

TomWerner committed:
EdgeWalker: fix the call to getMovementCost to properly handle reads/rivers

Using `neighbor` as both arguments means that river crossings and road usefulness was not always properly handled.

One of the added tests fails without this change, thinking that a move from <unroaded tile> to <roaded tile> only costs 1/3 of a movement point:

```
  Failed EngineTests.WalkerOnLandTest.testRoadOnDestinationNotOnStart [42 ms]
  Error Message:
   Assert.Contains() Failure
Not found: (filter expression)
In value:  List<Edge<Tile>> [Edge`1 { current = [0, 0] ( on ), distanceToCurrent = 0.33333334, prev = [0, 0] ( on ) }]
```

TomWerner committed:
EdgeWalker: fix the call to getMovementCost to properly handle reads/rivers

Using `neighbor` as both arguments means that river crossings and road usefulness was not always properly handled.

One of the added tests fails without this change, thinking that a move from <unroaded tile> to <roaded tile> only costs 1/3 of a movement point:

```
  Failed EngineTests.WalkerOnLandTest.testRoadOnDestinationNotOnStart [42 ms]
  Error Message:
   Assert.Contains() Failure
Not found: (filter expression)
In value:  List<Edge<Tile>> [Edge`1 { current = [0, 0] ( on ), distanceToCurrent = 0.33333334, prev = [0, 0] ( on ) }]
```

TomWerner committed:
MapUnitExtensions: Fix the movement cost for boats in cities

If we have a coastal city on a hill it doesn't make sense for it to take 2 movement points away from a boat.

One of the added tests fails without this change:

```
  Failed EngineTests.WalkerOnWaterTest.testLandIncludedIfItHasCity [214 ms]
  Error Message:
   Assert.Contains() Failure
Not found: (filter expression)
In value:  List<Edge<Tile>> [Edge`1 { current = [0, 0] ( on ), distanceToCurrent = 2, prev = [0, 0] ( on ) }]
```

TomWerner committed:
Uncomment line that was used for checking tests

TomWerner committed:
Introduce a new layer for displaying the "goto" cursor

This is a nice "quality of life" feature that also helped me identify
bugs in the movement cost calculations, such as the one fixed in [this
PR](#457).

This doesn't yet implement a line or some other mechanism to show the
actual planned path, but it both displays the tile that the user wants
to go to, and the cost of getting there in turns.

I moved the code for calculating the cost between two tiles into
`TilePath.cs`, since that's a lower dependency file, and this sort of
common utility should probably be in a low dependency file.

TomWerner committed:
removing accidentally added local change

TomWerner committed:
reverting comment change

TomWerner committed:
Add TODO for slight bug with path cost

esbudylin committed:
minor refactoring: UnitButtons

Node paths are moved from the source code into the scene file. It is
done using engine's export properties.

TomWerner committed:
Register the goto layer, this was accidentally reverted when dealing with some autoformatter changes

esbudylin committed:
Advisor menu fixes (#462)

- Exiting the advisor screen now correctly navigates to the main
screen.
- Game UI elements are no longer painted on top of the advisor screen
on small resolutions.

TomWerner committed:
Fix the issue with partial moves

TomWerner committed:
Draw the path a unit will take when using goto mode

This shows some instances where the pathing is a little goofy, but
usually is a nice visual represntation of what's going on.

I opted for the much simpler "dots on tiles" approach instead of trying
to draw lines representing the entire path, like civ3 does.

TomWerner committed:
Removed unnecessary if

WildWeazel committed:
Merge pull request #457 from TomWerner/godot4

EdgeWalker: fix the call to getMovementCost

WildWeazel committed:
Merge pull request #458 from TomWerner/fix-canal-movement-cost

MapUnitExtensions: Fix the movement cost for boats in cities

TomWerner committed:
Fixed the displayed move cost for moves that cost <1 turn, but aren't on railroads

TomWerner committed:
fixed merge

WildWeazel committed:
Merge pull request #459 from TomWerner/temp-work-on-move-cost

Introduce a new layer for displaying the "goto" cursor

TomWerner committed:
Fixed rounding issue for 4+ tile moves along roads, and ensured that the goto cursor is hidden for illegal moves

TomWerner committed:
Fixed rounding issue for 4+ tile moves along roads, and ensured that the goto cursor is hidden for illegal moves

WildWeazel committed:
Merge pull request #464 from TomWerner/show-path

Draw dots on the path the unit will take in "goto" mode

TomWerner committed:
Ensure that destroyed cities no longer show up in the UI (via ui message)

Now if a city is captured it is removed from the UI and the implicit
city road is removed.

To make this easier to test I modified `c7-static-map-save.json` locally
to add this:

```
    {
      "id": "Warrior-1",
      "prototype": "Warrior",
      "owner": "player-2",
      "previousLocation": {
        "x": -1,
        "y": -1
      },
      "currentLocation": {
        "x": 57,
        "y": 43
      },
      "hitPointsRemaining": 5,
      "movePointsRemaining": 1,
      "facingDirection": "north",
      "experience": "Elite"
    },
```

and modified this (to change the x/y position so an AI spawns next to
us, so our added warrior can destroy the city)

```
    {
      "id": "Settler-2",
      "prototype": "Settler",
      "owner": "player-3",
      "previousLocation": {
        "x": -1,
        "y": -1
      },
      "currentLocation": {
        "x": 59,
        "y": 43
      },
      "hitPointsRemaining": 3,
      "movePointsRemaining": 1,
      "facingDirection": "north",
      "experience": "Regular"
    },
```

TomWerner committed:
Hid some logic in CityLayer

TomWerner committed:
Make the retire button on the popup menu work

This now takes you back to the main menu, where you can start a new
game, retire, start a new game, etc.

TomWerner committed:
Merge pull request #470 from TomWerner/fix-retire-button

Make the retire button on the popup menu work

TomWerner committed:
Fix save game loading (#471)

This doesn't yet fix scenario loading; loading a scenario fails due to a
lack of a human player (`...does not contain a human player`).

Poking around in the history it looks like this was broken in [this
commit](https://github.com/C7-Game/Prototype/commit/caffcdd7eb65043938b27f34e1d04a3907d63733#diff-3fc4d4ac9715b2904cb97af15abca984009235b0f65f2da5cadf61fb6420c9edR22);
loading C7 saves would still work, but the conversion logic was
accidentally dropped.

TomWerner committed:
Util.cs: Properly check the C7 directory for art in debug builds

This ensures that the C7 settler art is actually used, instead of the
civ3 settler.

esbudylin committed:
Refactor node paths with export properties

Additionally, FileMode of SetCiv3HomeDialog is now configured via scene editor

esbudylin committed:
Merge remote-tracking branch 'origin/godot4' into export-properties

TomWerner committed:
Settler.ini: Fix the path format for the death sound (#467)

Prior to this the game would freeze, and the console would show errors
like `System.ApplicationException: Media path not found:
Art/Units/Settler/..\Worker\WorkerDeath.wav`.

esbudylin committed:
Implement player colors (#474)

Previous to this commit, player colors were handled incorrectly. The
color values loaded from the game files were represented as indexes
pointing to specific game assets. However, when drawing units and
cities, the game incorrectly used these indexes as RGB color
representation. As a result, every player appeared as black.

This commit introduces a utility method for loading player colors
from the game assets. This method transforms color indexes loaded from
a save file into proper RGB colors. This commit also modifies a part
of the code responsible for drawing units and cities on the game map.

To avoid confusion, which likely lead to the appearance of this
issue, the "color" field of Player, SavePlayer and Civilization
classes has been renamed to "colorIndex". The JSON save file has also
been changed accordingly.

I also removed placeholder game data from GameData class. It required
some changes since it used RGB colors instead of color indexes to
create players. However, since it no longer seems useful (dummy data
is now loaded from JSON instead of this class), I decided to remove
the code rather than maintaining it.

TomWerner committed:
globalize exactCaseRoot, which is much simpler

TomWerner committed:
revert whitespace

WildWeazel committed:
Merge pull request #463 from esbudylin/export-properties

[godot4] Manage node paths with export properties

WildWeazel committed:
Merge pull request #475 from TomWerner/fix-settler-animation-loading

Fix settler animation loading

TomWerner committed:
GameMap.cs: Move the tile neighbor calculations to Tile.cs (#472)

This puts this utility in a lower dependency module, which will allow me
to reuse it in a future PR that needs to calculate neighbors when
determining what tiles are visible for scenario loading.

TomWerner committed:
Fix scenario loading with no cities (#473)

* Fixed merge with civ color fix

* Swap if/for condition

TomWerner committed:
Set the default human player to carthage

The color index of 9 matches the entry for carthage on line 62747

WildWeazel committed:
436-carthage-credits (#466)

* Updating credits with new contributors and packages for Carthage, additional disclaimers, and version name

esbudylin committed:
Fix unresponsive unit buttons

Prior to this commit, some of the unit buttons (e.g., Build City,
Build Road) could become unresponsive. This issue was related to how
the game handled button presses. When a button was pressed, the game
emulated the press of an action hotkey, which should have been handled
by the Game class. However, in some cases, the input simulation didn't work
properly, and the game didn't react to the button presses.

To solve this issue, I decoupled the handling of button presses from
the handling of hotkey input. Now, instead of imitating input, a unit
button sends a signal to the Game class. In the Game class, the logic for
handling action requests has been moved into a new method, which is
now used for both handling signals from unit buttons and processing
hotkey input.

TomWerner committed:
When the last city of a civilization is destroyed, destroy the units too

... and display a popup so the user knows they accomplished this.

This gets us a step closer to being able to "win" a game of C7, instead
of just being able to produce units and fight.

TomWerner committed:
Clean up some comments

esbudylin committed:
Refactor: C7Action to TileDirection conversion

- Refactored the conversion from C7Action to TileDirection using a
switch expression. Moved the logic into the new method.
- Transformed fields of the C7Action class into constants to enable
their usage in the switch expression.

TomWerner committed:
Add TODO explaining more work is needed for full civ destruction logic

esbudylin committed:
Refactor: rewrite conversion method using nullable type

esbudylin committed:
Refactor handling of Godot signals

This commit addresses two issues with Godot signal handling, both of
which relate to the transition between engine versions. In Godot 3,
these were common practices for handling signals; however, the Godot 4
signal API offers more type-safe alternatives.

1) Use of raw strings for signal identification

Previously, raw strings were used to identify signals during
emission. This has been improved by using the nested SignalName
class provided by Godot signal API.

2) Use of the Connect() method for connecting signals

The Connect() method was used to connect methods to signals, a
practice discouraged by the Godot documentation. This commit
replaces Connect() with event mechanism.

WildWeazel committed:
correct action reference affected by merge #478
